src/Microsoft.Dnx.Watcher.Core/ProjectWatcher.cs

```diff
@@ -0,0 +1,286 @@
+            {
+                // TODO: better error message
+                throw new InvalidOperationException("Project file not found");
+            }
+
+            return projectOrDirectory;
+        }
+
+        public Task<bool> WatchAsync(string projectOrDirectory, string[] dnxArguments, string workingDir, CancellationToken cancellationToken)
+        {
+            _fullProjectFilePath = ResolveProjectFileToWatch(projectOrDirectory);
+            _rootPath = Path.GetDirectoryName(_fullProjectFilePath);
+
+            dnxArguments = new string[] { "--project", _fullProjectFilePath }
+               .Concat(dnxArguments)
+               .Select(arg =>
+               {
+                    // If the argument has spaces, make sure we quote it
+                    if (arg.Contains(" ") || arg.Contains("\t"))
+                   {
+                       return $"\"{arg}\"";
+                   }
+
+                   return arg;
+               })
+               .ToArray();
+
+            return WatchAsync(dnxArguments, workingDir, cancellationToken);
+        }
+
+        private async Task<bool> WatchAsync(string[] dnxArguments, string workingDir, CancellationToken cancellationToken)
+        {
+            var dnxArgumentsAsString = string.Join(" ", dnxArguments);
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                var project = await WaitForValidProjectJsonAsync(cancellationToken);
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return true;
+                }
+
+                using (var currentRunCancellationSource = new CancellationTokenSource())
+                using (var combinedCancellationSource = CancellationTokenSource.CreateLinkedTokenSource(
+                    cancellationToken,
+                    currentRunCancellationSource.Token))
+                {
+                    var fileWatchingTask = WaitForProjectFileToChangeAsync(project, combinedCancellationSource.Token);
+                    var dnxTask = WaitForDnxToExitAsync(dnxArgumentsAsString, workingDir, combinedCancellationSource.Token);
+
+                    var tasksToWait = new Task[] { dnxTask, fileWatchingTask };
+
+                    int finishedTaskIndex = Task.WaitAny(tasksToWait, cancellationToken);
+
+                    // Regardless of the outcome, make sure everything is cancelled
+                    // and wait for dnx to exit. We don't want orphan processes
+                    currentRunCancellationSource.Cancel();
+                    Task.WaitAll(tasksToWait);
+
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        return true;
+                    }
+
+                    if (finishedTaskIndex == 0)
+                    {
+                        // This is the dnx task
+                        var dnxExitCode = dnxTask.Result;
+
+                        if (dnxExitCode == 0)
```

Comment:
Actually that's not a leftover, just an incorrect write. It will write information when the exit code is 0 and error otherwise